api/v1/address.go

```diff
@@ -1,55 +1,104 @@
 package api
 
 import (
+	"encoding/json"
 	"net/http"
 
+	"github.com/data-preservation-programs/spade-tenant/db"
 	"github.com/labstack/echo/v4"
+	"gorm.io/gorm"
 )
 
-type Addresses []Address
-type AddressesMutable []AddressMutable
+type Addresses []db.Address
+type Address db.Address
 
-type Address struct {
-	AddressMutable
-	ActorID *uint `json:"actor_id,omitempty"` // TODO :swagger docs should be null not 0
+type AddressResponse struct {
+	AddressRobust    string `json:"address_robust" gorm:"primaryKey"`
+	AddressActorID   uint   `json:"actor_id"`
+	AddressIsSigning bool   `json:"is_signing" gorm:"default:true;not null"`
 }
 
-type AddressMutable struct {
-	Address   string `json:"address"`
-	IsSigning bool   `json:"is_signing"` // true - active dealmaking from this address, false - still counts as an associated wallet
+func ConfigureAddressesRouter(e *echo.Group, service *db.SpdTenantSvc) {
+	g := e.Group("/addresses")
+	g.PUT("", handleSetAddresses)
+	g.DELETE("", handleDeleteAddresses)
+	g.GET("", handleGetAddresses)
 }
 
 // handleSetAddresses godoc
 //
-//			@Summary		Update addresses associated with a tenant
-//	 		@Security apiKey
-//		  @Param 			addresses body AddressesMutable true "New addresses to add or change is_signing flag of"
-//			@Produce		json
-//			@Success		200	{object}	ResponseEnvelope{response=Addresses}
-//			@Router			/addresses [put]
+//	@Summary		Update addresses associated with a tenant
+//	@Param			token header string true "Auth token"
+//	@Param			addresses body AddressesMutable true "New addresses to add or change is_signing flag of"
+//	@Produce		json
+//	@Success		200	{object}	ResponseEnvelope{response=Addresses}
+//	@Router			/addresses [put]
 func handleSetAddresses(c echo.Context) error {
-	return c.JSON(http.StatusNotImplemented, map[string]string{})
+	var addresses Addresses
+	err := json.NewDecoder(c.Request().Body).Decode(&addresses)
+
+	if err != nil {
+		return c.JSON(http.StatusInternalServerError, CreateErrorResponseEnvelope(c, http.StatusInternalServerError, err.Error()))
+	}
+
+	err = db.DB.Transaction(func(tx *gorm.DB) error {
+		for _, address := range addresses {
+			address.TenantID = GetTenantId(c)
+			res := tx.Where("tenant_id = ? AND address_robust in (?) AND deleted_at IS NULL").Save(&address)
```

Comment:
this line seems a little strange, but did you test it and is it working as expected? Im not sure how `tx.Where()` will template in the values for `tenant_id` and `address_robust`. Also  `address_robust in` seems wrong, as you're looping over them you should just be checking against a single value 

I would think the "idiomatic Gorm" way to do this is like -
```go
db.DB.Model(&db.Address{}).Where("tenant_id = ?", address.TenantID).Where("address_robust = ?", address.AddressRobust).Updates(&address)
```

but needs a test to make sure it works - might be slightly different syntax wise

Reply:
Yep every call in the stack of diffs was tested with a local postgres instance. I believe it infers the address table from the save address but I can update it to be more gormy :P